README.md
# TI-99/Sim
a Texas Instruments Home Computer Emulator

# Introduction

This is the latest incarnation of an emulator for the TI-99/4A that I've been
working on, on and off, since 1993. At first it was a simple text-based
simulation of the TI (ti99sim-console) running under MS-DOS™. Then it was
ported to OS/2 and graphics support for the OS/2 Presentation Manager was
added. Later, it was ported to Windows and had sound support added. Now it's
been ported to Linux and Mac OS X and more features are still being added.

_DISCLAIMER_: I'm not one for writing a lot of documentation, so you're
encouraged to look at the code to see how some things work if they're not
described here. If you're interested in writing documentation, let me know.

_NOTE_: In order to run the emulator, you need to create a cartridge that
contains the console ROM & GROMs from the TI-99/4A. Texas Instruments will not
allow these to be distributed, so you'll need to find a way to get them
yourself. If you have a working copy of v9t9, instructions are included to
help you create the required cartridge. If you don't, I would suggest you
download it as it contains documentation on how to get the files you need to
do this.

# Getting Started

Before building or running TI-99/Sim you must install the SDL library. The
latest version is available at [http://www.libsdl.org/download-1.2.php](http:/
/www.libsdl.org/download-1.2.php). If you are planning on building TI-99/Sim,
you will need to install the Development Libraries version otherwise, the
Runtime Libraries package is sufficient.

## Building

If you downloaded the binary-only version, you can skip straight to
Installation. Before building you should make sure that the SDL development
code has been installed on your system and ensure that the sdl-config utility
is in you search path.

If everything is installed properly, running 'make' will build all the
executables found in the binary-only package. Any OS specific operations are
detailed in the following sections.

_NOTE_: When building from the source, all the executables are left in their
corresponding directories.

### Linux

Since this is the primary development environment, you should have few
problems building under Linux.

    
    
                **~/ti99sim>** make
              

### Macintosh OS X

In order to build ti99sim under Macintosh OS X, you'll need to install the
developer tools from Apple. You will also need to copy the SDL main files
(SDLMain.m and SDLMain.h) from the development libraries to the src/sdl
folder. Then, from a bash shell, just type:

    
    
                **~/ti99sim>** make
              

_NOTE_: I'm not a Mac OS X guru, so there is no neat little package for
TI-99/Sim. This probably means that to run it you'll need to do so from a
command window. (If anybody is interested in working on a proper Mac OS X
install is more than welcome to do so.)

### Windows

If you're building the win32 version, you'll need either a GNU-type
environment (GNU make, g++, ld, rm, ...) or Microsoft Visual C++.

If your using Microsoft Visual Studio, you can use the solution file found in
the vc9 subdirectory. These files assume that you have installed the SDL
development files in a folder called SDL-1.2.14 in a directory at the same
level as the TI-99/Sim directory.

The only GNU-type environment that I've tested is the mingw package, but
others may work as well. You will also need to make sure you define the WIN32
environment variable before building (type set WIN32=1 at the command prompt).
When you're ready, just type make in the directory where you placed the
emulator.

    
    
                **c:\ti99sim>** set WIN32=1
                **c:\ti99sim>** make
              

_NOTE_: If you are using mingw build tool, you will need to extract SDL_main.o
from libSDLmain.a (type: ar -x libSDLmain.a SDL_main.o) as well (until I can
figure out why libSDLmain.a isn't linking correctly with the mingw tools).

_NOTE_: The win32 command line build has not been tested in ages and may no
longer work correctly. Microsoft Visual Studio is the preferred development
environment for ti99sim on Windows.

## Installation

### The Directories

Installing the emulator is easy. Make sure you have sufficient privileges to
create/write to the installation directories and type:

    
    
                make install
              

This will create the default directories and install the binaries. The default
directory is /opt/ti99sim and the binaries will be installed in
/opt/ti99sim/bin. In addition to installing the binaries, symbolic links to
them will be created in /usr/local/bin (which should already be in your path).
When loading files (ROMs, cartridges, disk images, …), ti99sim will look in
the current directory, ~/.ti99sim, and finally, /opt/ti99sim. Under Windows,
the search is limited to the current directory followed by the directory
specified in the environment variable HOME (if it is defined at all). In each
of the directories searched, the following sub-directories may exist:

    
    
                ─┬─ cartridges
                 ├─ disks
                 └─ roms
              

When looking for a file, the appropriate sub-directory for that file type will
be searched. If there is no logical directory for a given file, only the base
directory will be searched.

### The Console ROM & Cartridges

Once you have run make install, you will need to create the console ROM
cartridge TI-994A.ctg (see convert-ctg below for instructions) in order to run
the emulator. This ROM cartridge contains the operating system, the BASIC
interpreter, and a description of the RAM configuration for the computer.
Without it, the emulator will not do anything useful. Once you have created
the console ROM cartridge, copy it to either ~/.ti99sim/roms or
/opt/ti99sim/roms.

You should also place a copy of spchrom.bin into this directory to enable
speech synthesis. If you don't have this file, you can use mkspch to create
one that will keep the emulator happy.

If you want to use the disk emulation features, you will need to create
another cartridge named ti-disk.ctg that contains a copy of the TI Disk
Controller ROM (the section on **convert-ctg** contains information about
creating cartridges for peripheral DSR ROMs).

A sample installation is shown below. In order to get up and running, however,
the only file needed is the TI-994A.ctg file.

    
    
                ─┬─ cartridges
                 │  ├─ Parsec.ctg
                 │  ├─ Alpiner.ctg
                 │  ├─ TI Extended BASIC.ctg
                 │  └─ Editor-Assembler.ctg
                 ├─ disks
                 │  ├─ ASSM1                   - E/A Disk 1
                 │  ├─ ASSM2                   - E/A Disk 2
                 │  ├─ DISKASSM                - Miller Graphics' Disk-Assembler
                 │  ├─ MG_DIAG                 - Miller Graphics' Advanced Diagnostics
                 │  └─ dsk1.dsk                - Default disk image
                 └─ roms
                    ├─ Gram Kracker.ctg        - Miller Graphic's Gram Kracker ROMs
                    ├─ TI-994A.ctg             - TI-99/4A ROM
                    ├─ spchrom.bin             - TI Speech Synthesizer ROM
                    └─ ti-disk.ctg             - TI ROM DSR module for disk access
              

### Up and Running

Now you're all set to run the emulator, just type:

    
    
                ti99sim-sdl
              

If everything is installed correctly, you should see a window on your screen
with the TI-99/4A startup screen followed by a beep. Enjoy!

Using the sample configuration in the previous section, the following are
valid commands:

    
    
                ti99sim-sdl --dsk2=ASSM1 --dsk3=ASSM2 Editor-Assembler.ctg   (DSK1 will default to using the dsk1.dsk image)
                ti99sim-sdl --dsk1=MG_DIAG "TI Extended BASIC.ctg"           (Note the use of quotes when using modules with spaces in their name)
                ti99sim-sdl -f Parsec.ctg                                    (OK, bring 'em on ...)
              

# Overview

## Disks

What good is a computer is you can't save information or programs? On the
TI-99/4A, there are two standard storage media: floppy disks and cassette
tapes. This section deals with floppy disks. Depending on your disk
controller, the TI could access either 3 or 4 floppy drives. These drive could
be either single-density (using FM encoding) or high-density (using MFM
encoding). Finally, disks can be either single-sided or double-sided. The
various disk sizes are listed in the following table.

TI-99/4A Disk Sizes & Nominal Values

Notation | Sides | Recording Mode | Sectors/Tracks | Tracks | Capacity
---------|-------|----------------|----------------|--------|---------
SSSD     |1      |FM              |9               |40      |90K
SSDD     |1      |MFM             |18              |40      |180K
DSSD     |2      |FM              |9               |80      |180K
DSDD     |2      |MFM             |18              |80      |360K
CF7+     |2      |CF*             |20              |80      |400K


doc/CHANGES.md
TI-99/Sim Changelog

Version 0.0.11

07-05-2012 - Fixed crashes in **disk**, **list**, and **convert-ctg**

07-02-2012 - Fixed logic for CapsLock

08-10-2011 - Added support for directories to **disk** utility

07-31-2011 - Fixed a crash in the **list** utility

07-30-2011 - Added support for CF7+ disk images (supported by both **disk**
and **list**)

07-25-2011 - Improved Mac OS X build process

03-05-2010 - Added support for using Windows .PIF files when creating
cartridges in **convert-ctg**

02-24-2010 - Fixed a bug in the TB opcode handling and the polarity of some
signals in the TMS9901 emulation**

02-18-2010 - Added support for REF/DEF table while loading most TI-Tagged
files in **dumpcpu**

02-17-2010 - Added support for most existing tags in TI-Tagged files in
**dumpcpu**

02-17-2010 - Fixed logic when reading variable size records (fixes bug in
**disk** when exporting and converting variable sized files)

11-30-2009 - Fixed timing issues when loading saved images and resetting the
CPU

09-18-2009 - Added 4K addressing mode support to TMS9918a emulation

08-10-2009 - Added checks for errors reported by file I/O functions

04-23-2009 - Fixed a crash when an invalid disk device cartridge is loaded

03-18-2009 - Fixed TimerHook in SDL build to handle fast systems properly &
track VDP refreshes more accurately

10-13-2008 - Updated Linux and Mac Makefile system

08-10-2004 - Added timestamp support to basic file system code

08-08-2004 - Added cross-linked file check option (--check) to **disk**
utility

07-29-2004 - Added --verbose output to **list** utility

07-28-2004 - Updated **list** utility to more closely match the output from TI
BASIC

06-22-2004 - Minor clean up - eliminated several lint/compiler warnings

01-05-2004 - Updated code to compile & run properly on 64-bit platforms

09-25-2003 - Updated cartridge format to use LZW compression

09-24-2003 - Added compression ratio information to verbose directory listing
for archive files

09-24-2003 - Modified LZW encoder & decoder to handle variable max bit-length
at run-time

09-23-2003 - Fixed bug in cDecodeLZW that prevented some valid files from
decoding the last symbol properly

09-23-2003 - Fixed cDecodeLZW's handling of corrupt data

09-22-2003 - Added cEncodeLZW encoder class

Version 0.0.10

09-15-2003 - Added cArchiveFileSystem class to handle .ark files as disks

09-15-2003 - Updated list of files to strip/install under Win32

09-14-2003 - Wrote LZW routines to decode .ark files

09-10-2003 - Added support for REF records in TI-Tagged files to **dumpcpu**

09-01-2003 - Added ability to add files to disk images in **disk**

Version 0.0.9

08-29-2003 - Fix **convert-ctg**'s handling of Gram Kracker files (10
character filename limit & PC99 formats)

08-03-2003 - Add 'Scale3x' code and optimized 'Scale2x' routines

05-01-2003 - Improved quality of speech synthesis when using External Data
mode

01-16-2003 - Fixed bug in **list** utility (failing to NULL terminate the
string)

12-05-2002 - Added 'volume' and 'scale2x' command-line option

12-02-2002 - Added option to select an alternate color palette

12-01-2002 - Fixed window down-scaling

12-01-2002 - Implemented Scale2x code for double-size windows

11-21-2002 - Changed joystick code to handle X/Y mappings better

11-22-2002 - Changed sound code to use unsigned samples

11-24-2002 - Added more interpolation code to speech synthesis code to improve
quality (slightly)

11-13-2002 - Fixed command line parsing to notify user of invalid arguments

11-13-2002 - Added diagnostic messages to ParseJoystick in the case of bad
input

09-17-2002 - Added window resizing support to **ti99sim-sdl**

09-17-2002 - Added new fullscreen option to select a specific resolution

09-17-2002 - Added better auto-selection of screen resolution in fullscreen
mode

09-17-2002 - Changed cBitmap::Copy to handle arbitrary integer upscaling

09-15-2002 - Modified startup code for **ti99sim-sdl** to handle platforms
that don't support joysticks properly

09-15-2002 - Fixed code in decode.cpp to byte-swap structure members on big
endian processors

09-15-2002 - Reworked Makefile.linux to simplify adding new utilities, added
uninstall target, and added soft links in /usr/local/bin

09-14-2002 - Ported code to run on PowerPC platform (MacOSX)

09-14-2002 - Removed bitfields from code & replaced with masks and shifts

09-14-2002 - Fixed display problem in **ti99sim-console**

Version 0.0.8 - 20-Aug-2002

08-03-2002 - Fixed bug in **mkspch** that created incorrect binary tree

08-03-2002 - Added parsed data mode to **dumpspch** and **mkspch**

07-31-2002 - Created **dumpspch** utility

07-31-2002 - Fixed several memory leaks/invalid memory accesses in utilities


src/sdl/main.cpp
//----------------------------------------------------------------------------
//
// File:        main.cpp
// Date:        02-Apr-2000
// Programmer:  Marc Rousseau
//
// Description: This file contains starup code for Linux/SDL
//
// Copyright (c) 2000-2004 Marc Rousseau, All Rights Reserved.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
//
// Revision History:
//
//----------------------------------------------------------------------------

#include <stdio.h>
#include "common.hpp"
#include "logger.hpp"
#include "cartridge.hpp"
#include "SDL.h"
#include "tms9900.hpp"
#include "ti994a.hpp"
#include "ti994a-sdl.hpp"
#include "tms9918a.hpp"
#include "tms9918a-sdl.hpp"
#include "tms9919.hpp"
#include "tms9919-sdl.hpp"
#include "tms5220.hpp"
#include "device.hpp"
#include "diskio.hpp"
#include "ti-disk.hpp"
#include "support.hpp"
#include "option.hpp"

DBG_REGISTER ( __FILE__ );

static sRGBQUAD ColorTable [3][17] = {
    {
        { 0x00, 0x00, 0x00, 0x00 },		// 0x00 TI_TRANSPARENT - Background
        { 0x00, 0x00, 0x00, 0x00 },		// 0x01 TI_BLACK
        { 0x48, 0x9C, 0x08, 0x00 },		// 0x02 TI_MEDIUM_GREEN *
        { 0x70, 0xBF, 0x88, 0x00 },		// 0x03 TI_LIGHT_GREEN *
        { 0x28, 0x3C, 0x8A, 0x00 },		// 0x04 TI_DARK_BLUE *
        { 0x50, 0x6C, 0xCF, 0x00 },		// 0x05 TI_LIGHT_BLUE *
        { 0xD0, 0x48, 0x00, 0x00 },		// 0x06 TI_DARK_RED *
        { 0x00, 0xCC, 0xFF, 0x00 },		// 0x07 TI_CYAN *
        { 0xD0, 0x58, 0x28, 0x00 },		// 0x08 TI_MEDIUM_RED *
        { 0xFF, 0xA0, 0x40, 0x00 },		// 0x09 TI_LIGHT_RED *
        { 0xFC, 0xF0, 0x50, 0x00 },		// 0x0A TI_DARK_YELLOW *
        { 0xFF, 0xFF, 0x80, 0x00 },		// 0x0B TI_LIGHT_YELLOW *
        { 0x00, 0x80, 0x00, 0x00 },		// 0x0C TI_DARK_GREEN
        { 0xCD, 0x58, 0xCD, 0x00 },		// 0x0D TI_MAGENTA *
        { 0xE0, 0xE0, 0xE0, 0x00 },		// 0x0E TI_GRAY *
        { 0xFF, 0xFF, 0xFF, 0x00 },		// 0x0F TI_WHITE
        { 0xFF, 0xFF, 0xFF, 0x00 }		// 0x10 - Text Mode Foreground
    }, {
        { 0x00, 0x00, 0x00, 0x00 },		// 0x00 TI_TRANSPARENT - Background
        { 0x00, 0x00, 0x00, 0x00 },		// 0x01 TI_BLACK
        {  33, 200,  66, 0x00 },		// 0x02 TI_MEDIUM_GREEN *
        {  94, 220, 120, 0x00 },		// 0x03 TI_LIGHT_GREEN *
        {  84,  85, 237, 0x00 },		// 0x04 TI_DARK_BLUE *
        { 125, 118, 252, 0x00 },		// 0x05 TI_LIGHT_BLUE *
        { 212,  82,  77, 0x00 },		// 0x06 TI_DARK_RED *
        {  66, 235, 245, 0x00 },		// 0x07 TI_CYAN *
        { 252,  85,  84, 0x00 },		// 0x08 TI_MEDIUM_RED *
        { 255, 121, 120, 0x00 },		// 0x09 TI_LIGHT_RED *
        { 212, 193,  84, 0x00 },		// 0x0A TI_DARK_YELLOW *
        { 230, 206, 128, 0x00 },		// 0x0B TI_LIGHT_YELLOW *
        {  33, 176,  59, 0x00 },		// 0x0C TI_DARK_GREEN
        { 201,  91, 186, 0x00 },		// 0x0D TI_MAGENTA *
        { 204, 204, 204, 0x00 },		// 0x0E TI_GRAY *
        { 0xFF, 0xFF, 0xFF, 0x00 },		// 0x0F TI_WHITE
        { 0xFF, 0xFF, 0xFF, 0x00 }		// 0x10 - Text Mode Foreground
    }, {
        { 0x00, 0x00, 0x00, 0x00 },		// 0x00 TI_TRANSPARENT - Background
        { 0x00, 0x00, 0x00, 0x00 },		// 0x01 TI_BLACK
        { 0x00, 0xCC, 0x00, 0x00 },		// 0x02 TI_MEDIUM_GREEN
        { 0x00, 0xFF, 0x00, 0x00 },		// 0x03 TI_LIGHT_GREEN
        { 0x00, 0x00, 0x80, 0x00 },		// 0x04 TI_DARK_BLUE
        { 0x00, 0x00, 0xFF, 0x00 },		// 0x05 TI_LIGHT_BLUE
        { 0x80, 0x00, 0x00, 0x00 },		// 0x06 TI_DARK_RED
        { 0x00, 0xFF, 0xFF, 0x00 },		// 0x07 TI_CYAN
        { 0xCC, 0x00, 0x00, 0x00 },		// 0x08 TI_MEDIUM_RED
        { 0xFF, 0x00, 0x00, 0x00 },		// 0x09 TI_LIGHT_RED
        { 0xB0, 0xB0, 0x00, 0x00 },		// 0x0A TI_DARK_YELLOW
        { 0xFF, 0xFF, 0x00, 0x00 },		// 0x0B TI_LIGHT_YELLOW
        { 0x00, 0x80, 0x00, 0x00 },		// 0x0C TI_DARK_GREEN
        { 0xB0, 0x00, 0xB0, 0x00 },		// 0x0D TI_MAGENTA
        { 0xCC, 0xCC, 0xCC, 0x00 },		// 0x0E TI_GRAY
        { 0xFF, 0xFF, 0xFF, 0x00 },		// 0x0F TI_WHITE
        { 0xFF, 0xFF, 0xFF, 0x00 }		// 0x10 - Text Mode Foreground
    }
};

static int   geometryX            = VDP_WIDTH;
static int   geometryY            = VDP_HEIGHT;
static bool  joystickInitialized  = false;
static int   joystickIndex [2]    = { 0, 1 };
static int   framesOn             = 1;
static int   framesOff            = 0;
static char *diskImage [3];

bool ListJoysticks ( const char *, void * )
{
    FUNCTION_ENTRY ( NULL, "ListJoysticks", true );

    if (( joystickInitialized == false ) && ( SDL_InitSubSystem ( SDL_INIT_JOYSTICK ) < 0 )) {
        fprintf ( stderr, "Error initializing joystick subsystem: %s\n", SDL_GetError ());
        return false;
    }

    joystickInitialized = true;

    int num = SDL_NumJoysticks ();
    fprintf ( stdout, ( num == 1 ) ? "1 joystick was found.\n" : "%i joysticks were found.\n", num );

    if ( num > 0 ) fprintf ( stdout, "\nThe names of the joysticks are:\n" );
    for ( int i = 0; i < SDL_NumJoysticks (); i++ ) {
        fprintf ( stdout, "  %d) %s\n", i + 1, SDL_JoystickName ( i ));
    }

    exit ( 0 );
}

bool ListResolutions ( const char *, void * )
{
    FUNCTION_ENTRY ( NULL, "ListResolutions", true );

    const int MAX_RESOLUTIONS = 32;

    int xDisplay [MAX_RESOLUTIONS];
    int yDisplay [MAX_RESOLUTIONS];

    int n = cSdlTMS9918A::GetFullScreenResolutions ( xDisplay, yDisplay, MAX_RESOLUTIONS );

    fprintf ( stdout, "Available full screen resolutions:\n" );

    for ( int i = 0; i < n; i++ ) {
        fprintf ( stdout, "%2d - %4dx%d\n", i + 1, xDisplay [i], yDisplay [i] );
    }

    exit ( 0 );
}

bool ParseJoystick ( const char *arg, void * )
{
    FUNCTION_ENTRY ( NULL, "ParseJoystick", true );

    arg += strlen ( "joystick" );

    int joy_num = arg [0] - '1';
    if (( joy_num < 0 ) || ( joy_num > 1 ) || ( arg [1] != '=' )) {
        fprintf ( stderr, "Joystick must be either 1 or 2\n" );
        return false;
    }

    if (( joystickInitialized == false ) && ( SDL_InitSubSystem ( SDL_INIT_JOYSTICK ) < 0 )) {
        fprintf ( stderr, "Error initializing joystick subsystem: %s\n", SDL_GetError ());
        return false;
    }

    joystickInitialized = true;

    int joy_index = atoi ( arg + 2 ) - 1;
    if (( joy_index < 0 ) || ( joy_index >= SDL_NumJoysticks ())) {
        if ( SDL_NumJoysticks () == 0 ) {
            fprintf ( stderr, "No joysticks were detected\n" );
        } else {
            fprintf ( stderr, "Joystick %s is invalid (range=%d-%d)\n", arg + 2, 1, SDL_NumJoysticks ());
        }
        return false;
    }
    joystickIndex [joy_num] = joy_index;

    return true;
}

bool ParseDisk ( const char *arg, void * )
{
    FUNCTION_ENTRY ( NULL, "ParseDisk", true );

    arg += strlen ( "dsk" );

    int disk = arg [0] - '1';
    if (( disk < 0 ) || ( disk > 2 ) || ( arg [1] != '=' )) {
        fprintf ( stderr, "Disk must be either 1, 2, or 3\n" );
        return false;
    }

    diskImage [disk] = strdup ( arg + 2 );

    return true;
}

bool ParseSampleRate ( const char *arg, void *ptr )
{
    FUNCTION_ENTRY ( NULL, "ParseSampleRate", true );

    int freq = 0;

    arg = strchr ( arg, '=' ) + 1;

    if ( sscanf ( arg, "%d", &freq ) != 1 ) {
        fprintf ( stderr, "Invalid sampling rate '%s'\n", arg );
        return false;
    }

    if (( freq > 44100 ) || ( freq < 8000 )) {
        fprintf ( stderr, "Sampling rate must be between 8000 and 44100\n" );
        return false;
    }

    * ( int * ) ptr = freq;

    return true;
}

bool ParseFrameRate ( const char *arg, void * )
{
    FUNCTION_ENTRY ( NULL, "ParseFrameRate", true );

    int num = 0, den = 0;

    arg = strchr ( arg, '=' ) + 1;

    switch ( sscanf ( arg, "%d/%d", &num, &den )) {
        case 1 :
            den = 100;
            break;
        case 2 :
            break;
        default :
            fprintf ( stderr, "Invalid framerate '%s'\n", arg );
            return false;
    }

    if (( num > den ) || ( num * den == 0 )) {
        fprintf ( stderr, "Invalid framerate specified\n" );
        return false;
    }

    framesOn  = num;
    framesOff = den - num;

    return true;
}

bool IsType ( const char *filename, const char *type )
{
    FUNCTION_ENTRY ( NULL, "IsType", true );

    size_t len = strlen ( filename );
    const char *ptr = filename + len - 4;
    return ( strcmp ( ptr, type ) == 0 ) ? true : false;
}

void PrintUsage ()
{
    FUNCTION_ENTRY ( NULL, "PrintUsage", true );

    fprintf ( stdout, "Usage: ti99sim-sdl [options] [cartridge.ctg] [image.img]\n" );
    fprintf ( stdout, "\n" );
}

int main ( int argc, char *argv [] )
{
    FUNCTION_ENTRY ( NULL, "main", true );

    int  flagSize        = 1;
    bool flagSound       = true;
    bool flagSpeech      = true;
    bool flagJoystick    = true;
    int  colorTableIndex = 0;
    int  fullScreenMode  = -1;
    int  refreshRate     = 60;
    int  samplingRate    = 44100;
    bool useScale2x      = false;
    int  volume          = 50;

    sOption optList [] = {
        { '4', NULL,                 OPT_VALUE_SET | OPT_SIZE_INT,  2,     &flagSize,        NULL,            "Double width/height window" },
        {  0,  "dsk*n=<filename>",   OPT_NONE,                      0,     NULL,             ParseDisk,       "Use <filename> disk image for DSKn" },
        {  0,  "framerate=*{n/d|p}", OPT_NONE,                      0,     NULL,             ParseFrameRate,  "Reduce frame rate to fraction n/d or percentage p" },
        { 'f', "fullscreen*=n",      OPT_VALUE_PARSE_INT,           0,     &fullScreenMode,  NULL,            "Fullscreen" },
        {  0,  "joystick*n=i",       OPT_NONE,                      0,     NULL,             ParseJoystick,   "Use system joystick i as TI joystick n" },
        {  0,  "list-joysticks",     OPT_NONE,                      0,     NULL,             ListJoysticks,   "Print a list of all detected joysticks" },
        {  0,  "list-resolutions",   OPT_NONE,                      0,     NULL,             ListResolutions, "Print a list of available fullscreen resolutions" },
        {  0,  "no-joystick",        OPT_VALUE_SET | OPT_SIZE_BOOL, false, &flagJoystick,    NULL,            "Disable hardware joystick support" },
        { 'q', "no-sound",           OPT_VALUE_SET | OPT_SIZE_BOOL, false, &flagSound,       NULL,            "Turn off all sound/speech" },
        {  0,  "no-speech",          OPT_VALUE_SET | OPT_SIZE_BOOL, false, &flagSpeech,      NULL,            "Disable speech synthesis" },
        {  0,  "NTSC",               OPT_VALUE_SET | OPT_SIZE_INT,  60,    &refreshRate,     NULL,            "Emulate a NTSC display (60Hz)" },
        {  0,  "PAL",                OPT_VALUE_SET | OPT_SIZE_INT,  50,    &refreshRate,     NULL,            "Emulate a PAL display (50Hz)" },
        { 'p', "palette=*n",         OPT_VALUE_PARSE_INT,           0,     &colorTableIndex, NULL,            "Select a color palette (1-3)" },
        { 's', "sample=*<freq>",     OPT_NONE,                      0,     &samplingRate,    ParseSampleRate, "Select sampling frequency for audio playback" },
        {  0,  "scale2x",            OPT_VALUE_SET | OPT_SIZE_BOOL, true,  &useScale2x,      NULL,            "Use the Scale2x algorithm to scale display" },
        { 'v', "verbose*=n",         OPT_VALUE_PARSE_INT,           1,     &verbose,         NULL,            "Display extra information" },
        {  0,  "volume=*n",          OPT_VALUE_PARSE_INT,           50,    &volume,          NULL,            "Set the audio volume" }
    };

    // Initialize the SDL library (starts the event loop)
    if ( SDL_Init ( SDL_INIT_NOPARACHUTE | SDL_INIT_VIDEO | SDL_INIT_AUDIO ) < 0 ) {
        fprintf ( stderr, "Couldn't initialize SDL: %s\n", SDL_GetError ());
        return -1;
    }

    // Clean up on exit, exit on window close and interrupt
    atexit ( SDL_Quit );

    SDL_WM_SetCaption ( "TI-99/sim", NULL );

    SDL_EnableUNICODE ( 1 );

    const char *ctgFile = NULL;
    const char *imgFile = NULL;

    int index = 1;

    bool isValid = true;

    while ( index < argc ) {

        int newIndex = ParseArgs ( index, argc, argv, SIZE ( optList ), optList );

        if ( index == newIndex ) {

            if ( IsType ( argv [index], ".ctg" )) {

                const char *filename = LocateFile ( argv [index], "cartridges" );
                if ( filename != NULL ) {
                    if ( ctgFile != NULL ) free (( void * ) ctgFile );
                    ctgFile = strdup ( filename );
                } else {
                    isValid = false;
                    fprintf ( stderr, "Unable to locate cartridge \"%s\"\n", argv [index] );
                }

            } else if ( IsType ( argv [index], ".img" )) {

                const char *filename = LocateFile ( argv [index] );
                if ( filename != NULL ) {
                    isValid = true;
                    if ( imgFile != NULL ) free (( void * ) imgFile );
                    imgFile = strdup ( filename );
                } else {
                    isValid = false;
                    fprintf ( stderr, "Unable to locate image \"%s\"\n", argv [index] );
                }

            } else {

                isValid = false;
                fprintf ( stderr, "Unrecognized argument \"%s\"\n", argv [index] );

            }


            index++;

        } else {

            index = newIndex;

        }
    }

    if ( isValid == false ) {
        return 0;
    }

    if ( colorTableIndex > 0 ) {
        if ( colorTableIndex > 3 ) {
            fprintf ( stderr, "Invalid palette selected - must be 1 or 2\n" );
            return -1;
        }
        colorTableIndex--;
    }

    SDL_Joystick *joy1 = NULL;
    SDL_Joystick *joy2 = NULL;

    if ( flagJoystick == true ) {
        // Try to initialize the joystick subsystem and allocate them
        if (( joystickInitialized == true ) || ( SDL_InitSubSystem ( SDL_INIT_JOYSTICK ) >= 0 )) {
            // NOTE: If we called SDL_InitSubSystem here, the user didn't explicitly request joystick support
            //   and we'll silently ignore the failure and disable joysticks (MAC OS X seems to have a
            //   problem with joysticks).  If they asked for them, and it failed it would have been reported
            //   to them in ParseJoystick.
            joystickInitialized = true;
            joy1 = SDL_JoystickOpen ( joystickIndex [0] );
            joy2 = SDL_JoystickOpen ( joystickIndex [1] );
        }
    }

    const char *romFile = LocateFile ( "TI-994A.ctg", "roms" );
    if ( romFile == NULL ) {
        fprintf ( stderr, "Unable to locate console ROMs!\n" );
        return -1;
    }

    if ( verbose > 0 ) fprintf ( stdout, "Using system ROM \"%s\"\n", romFile );
    cCartridge *consoleROM = new cCartridge ( romFile );

    if ( flagSize == 2 ) {
        geometryX *= 2;
        geometryY *= 2;
    }

    if ( fullScreenMode != -1 ) {
        const int MAX_RESOLUTIONS = 32;
        int xDisplay [MAX_RESOLUTIONS];
        int yDisplay [MAX_RESOLUTIONS];
        int n = cSdlTMS9918A::GetFullScreenResolutions ( xDisplay, yDisplay, MAX_RESOLUTIONS );
        if (( fullScreenMode > 0 ) && ( fullScreenMode <= n )) {
            geometryX = xDisplay [fullScreenMode-1];
            geometryY = yDisplay [fullScreenMode-1];
        } else {
            geometryX = 0;
            geometryY = 0;
        }
    }

    if ( flagSound == ( int ) false ) flagSpeech = false;

    cSdlTMS9918A *vdp = new cSdlTMS9918A ( ColorTable [colorTableIndex], refreshRate, useScale2x, ( fullScreenMode != -1 ) ? true : false, geometryX, geometryY );
    cTMS9919 *sound = NULL;
    if ( flagSound == false ) {
        sound = new cTMS9919 ();
    } else {
        cSdlTMS9919 *sdlSound = new cSdlTMS9919 ( samplingRate );
        sdlSound->SetMasterVolume ( volume );
        sound = sdlSound;
    }
    cTMS5220 *speech = ( flagSpeech == false ) ? NULL : new cTMS5220 ( sound );

    vdp->SetFrameRate ( framesOn, framesOff );

    cSdlTI994A computer ( consoleROM, vdp, sound, speech );

    const char *diskFile = LocateFile ( "ti-disk.ctg", "roms" );
    if ( diskFile != NULL ) {
        if ( verbose > 0 ) fprintf ( stdout, "Loading disk ROM \"%s\"\n", diskFile );
        cDiskDevice *disk = new cDiskDevice ( diskFile );
        // Look for DSKn images
        for ( unsigned i = 0; i < SIZE ( diskImage ); i++ ) {
            char buffer [256];
            const char *validName = LocateFile ( diskImage [i], "disks" );
            if ( validName == NULL ) {
                char dskName [10];
                sprintf ( dskName, "dsk%d.dsk", i + 1 );
                validName = LocateFile ( dskName, "disks" );
                if ( validName == NULL ) {
                    // Can't find one - use ~/.ti99sim/disks/dskn.dsk as the default
                    sprintf ( buffer, "%s%c%s%c%s", HOME_PATH, FILE_SEPERATOR, "disks", FILE_SEPERATOR, dskName );
                    validName = buffer;
                }
            }
            if ( verbose > 0 ) fprintf ( stdout, "  DSK%d: \"%s\"\n", i+1, validName );
            disk->LoadDisk ( i, validName );
        }
        computer.AddDevice ( disk );
    } else {
        if ( verbose > 0 ) fprintf ( stdout, "Unable to locate disk ROM images \"%s\"\n", "ti-disk.ctg" );
    }

    if ( joy1 != NULL ) computer.SetJoystick ( 0, joy1 );
    if ( joy2 != NULL ) computer.SetJoystick ( 1, joy2 );

    cCartridge *ctg = NULL;
    if ( ctgFile != NULL ) {
        ctg = new cCartridge ( ctgFile );
        if ( verbose > 0 ) fprintf ( stdout, "Loading cartridge \"%s\" (%s)\n", ctgFile, ctg->Title ());
        computer.InsertCartridge ( ctg );
    }

    if ( imgFile != NULL ) {
        if ( verbose > 0 ) fprintf ( stdout, "Loading image \"%s\"\n", imgFile );
        (( cTI994A & ) computer).LoadImage ( imgFile );
    }

    if ( verbose > 0 ) {
        fprintf ( stdout, " Video refresh rate: %d Hz\n", refreshRate );
        if ( flagSound == true ) fprintf ( stdout, "Audio sampling rate: %d Hz\n", samplingRate );
    }

    computer.Run ();

    if ( ctg != NULL ) {
        computer.RemoveCartridge ( ctg );
        delete ctg;
    }

    for ( unsigned i = 0; i < SIZE ( diskImage ); i++ ) {
        if ( diskImage [i] != NULL ) {
            free ( diskImage [i] );
        }
    }

    if ( SDL_JoystickOpened ( joystickIndex [0] )) SDL_JoystickClose ( joy1 );
    if ( SDL_JoystickOpened ( joystickIndex [1] )) SDL_JoystickClose ( joy2 );

    if ( ctgFile != NULL ) free (( void * ) ctgFile );
    if ( imgFile != NULL ) free (( void * ) imgFile );

    return 0;
}
